Firestore/core/src/local/leveldb_persistence.h

```diff
@@ -123,11 +134,15 @@ class LevelDbPersistence : public Persistence {
   bool started_ = false;
 
   std::unique_ptr<LevelDbBundleCache> bundle_cache_;
-  std::unique_ptr<LevelDbDocumentOverlayCache> current_document_overlay_cache_;
-  std::unique_ptr<LevelDbMutationQueue> current_mutation_queue_;
+  std::unordered_map<std::string, std::unique_ptr<LevelDbDocumentOverlayCache>>
+      document_overlay_caches_;
+  std::unique_ptr<LevelDbOverlayMigrationManager> overlay_migration_manager_;
+  std::unordered_map<std::string, std::unique_ptr<LevelDbMutationQueue>>
+      mutation_queues_;
   std::unique_ptr<LevelDbTargetCache> target_cache_;
   std::unique_ptr<LevelDbRemoteDocumentCache> document_cache_;
-  std::unique_ptr<LevelDbIndexManager> index_manager_;
+  std::unordered_map<std::string, std::unique_ptr<LevelDbIndexManager>>
+      index_managers_;
```

Comment:
IIUC you're keeping these 4 tables in memory for all users because the overlay migration code is iterating over all users. Should we be concerned about the amount of data that's kept in memory here?

IIUC the caching done here is not necessary, but a performance improvement _for only when overlay migration happens_. OTOH this will increase the memory footprint for all use cases when user switches. Is my understanding correct? Is it worth to take the performance hit during the overlay migration, but not take a memory hit for regular operations?

Reply:
What prompted me to make this change is: persistence handles out a bunch of raw pointers to document_overlay_cache, mutation_queue, etc. When MigrationManager enumerate users and call `GetDocumentOverlayCache(user)`, it invalidates the previous unique_ptr, and leaves local_store holding dangling pointers.

This is not a problem with Java because of GC, the change I made basically is a GC simulation (where the collection happens when SDK exits..). I can add more logic to do this when migration is done however. Did not think it's worth it but now it seems necessary.

Reply:
SGTM. Thanks